File: UITextSubClass/UITextFieldWithDecimalPad/UITextFieldWithDecimalPad.m
//
//  UITextFieldWithDecimalPad.m
//  MotherChildHandbook
//
//  Created by P.I.akura on 2013/09/30.
//  Copyright (c) 2013年 P.I.akura. All rights reserved.
//

#import "UITextFieldWithDecimalPad.h"

@implementation UITextFieldWithDecimalPad

- (id)initWithCoder:(NSCoder *)aDecoder {
    self = [super initWithCoder:aDecoder];
    if (self) {
        self.keyboardType = UIKeyboardTypeDecimalPad;
        self.menuHidden = YES;
        [self addTarget:self action:@selector(didChangeText:) forControlEvents:UIControlEventEditingChanged];
    }
    return self;
}

- (void)didChangeText:(id)sender {
    NSString *phoneNumberRegex = @"^([1-9]\\d*|0)?(\\.(\\d+)?)?$";
    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"SELF MATCHES %@", phoneNumberRegex];
    if ([predicate evaluateWithObject:self.text] == NO) {
        self.value = self.text.floatValue;
    }
}

- (CGFloat)value {
    return [self format:[self.text floatValue]];
}

- (void)setValue:(CGFloat)value {
    super.text = [@([self format:value]) stringValue];
}

- (CGFloat)format:(CGFloat)value {
    NSInteger num = pow(10, self.significantFigures);
    return ((NSInteger)(value * num))/(CGFloat)num;
}

- (BOOL)canPerformAction:(SEL)action withSender:(id)sender {
    if (self.menuHidden) {
        return NO;
    } else {
        return YES;
    }
}

- (NSUndoManager *)undoManager {
    return nil;
}

@end

File: UITextSubClass/UITextFieldWithDecimalPad/UITextFieldWithDecimalPad.swift
import UIKit

public class UITextFieldWithDecimalPad: UITextField {
    public var value: NSDecimalNumber {
        get { format(number: .init(string: text)) }
        set { text = format(number: newValue).stringValue }
    }
    public var handler: NSDecimalNumberHandler

    public required init?(coder: NSCoder) {
        handler = .default
        super.init(coder: coder)
        setup()
    }

    public init(handler: NSDecimalNumberHandler) {
        self.handler = handler
        super.init(frame: .zero)
        setup()
    }

    func setup() {
        value = .zero
        keyboardType = .decimalPad
        addTarget(self, action: #selector(editingChanged), for: .editingChanged)
    }

    @objc func editingChanged(_ sender: UITextField) {
        guard shouldBeReformat(text: text ?? "") else {
            return
        }
        text = format(number: .init(string: text)).stringValue
        undoManager?.removeAllActions()
    }
    func shouldBeReformat(text: String) -> Bool {
        let phoneNumberRegex = "^([1-9]\\d*|0)?(\\.(\\d+)?)?$"
        let predicate = NSPredicate(format: "SELF MATCHES %@", phoneNumberRegex)
        if predicate.evaluate(with: text) {
            let number = NSDecimalNumber(string: text)
            let formatNumber = format(number: number)
            if number == formatNumber {
                return false
            } else {
                return true
            }
        } else {
            return true
        }
    }
    func format(number: NSDecimalNumber) -> NSDecimalNumber {
        number.adding(.zero, withBehavior: handler)
    }
}
